File: CalibrationLoader.swift
//
// RWPData.swift
// ARMetrologyApp - Configuration Data Model
//

import simd

/// A structure for deserializing Real-World Physical (RWP) coordinates
/// from a configuration file (RWPConfig.json).
struct RWPData: Decodable {
    let id: String
    let x: Float
    let y: Float
    let z: Float
    
    // Helper to convert to a SIMD3<Float>
    var coordinate: SIMD3<Float> {
        return SIMD3<Float>(x, y, z)
    }
}

/// Root structure for the RWP configuration file, holding a list of points.
struct RWPConfiguration: Decodable {
    let points: [RWPData]
}

//
// CalibrationLoader.swift
// ARMetrologyApp - Utility Layer
//

import Foundation

/// Handles loading and parsing of the Real-World Physical (RWP) reference points
/// from the RWPConfig.json file, making the app dynamically configurable.
class CalibrationLoader {
    
    /// Loads RWP points from a JSON file in the main bundle and populates the CalibrationManager.
    /// - Parameters:
    ///   - manager: The CalibrationManager instance to populate.
    ///   - fileName: The name of the JSON file (e.g., "RWPConfig").
    /// - Returns: An ordered list of point IDs.
    /// - Throws: An error if the file cannot be loaded or parsed.
    static func loadRwpPoints(into manager: CalibrationManager, from fileName: String = "RWPConfig") throws -> [String] {
        guard let url = Bundle.main.url(forResource: fileName, withExtension: "json") else {
            throw CalibrationError.custom(message: "RWP Configuration file '\(fileName).json' not found in bundle.")
        }
        
        let data = try Data(contentsOf: url)
        let decoder = JSONDecoder()
        let config = try decoder.decode(RWPConfiguration.self, from: data)
        
        guard !config.points.isEmpty else {
            throw CalibrationError.custom(message: "RWP configuration contains no points.")
        }
        
        var orderedPointIds: [String] = []
        
        for rwpData in config.points {
            manager.addRealWorldReferencePoint(id: rwpData.id, coordinate: rwpData.coordinate)
            orderedPointIds.append(rwpData.id)
        }
        
        return orderedPointIds
    }
}

File: CalibrationLoader.cs
//
// CalibrationLoader.cs
// MetrologyCalibration - Utility Layer
//

using System;
using System.IO;
using System.Text.Json;
using System.Collections.Generic;

/// <summary>
/// Handles loading and parsing of the Real-World Physical (RWP) reference points
/// from the RWPConfig.json file.
/// </summary>
public static class CalibrationLoader
{
    /// <summary>
    /// Loads RWP points from a JSON file and populates the CalibrationManager.
    /// </summary>
    /// <param name="manager">The CalibrationManager instance to populate.</param>
    /// <param name="jsonFilePath">The file path to the RWPConfig.json.</param>
    /// <returns>A list of point IDs in the order they were defined.</returns>
    /// <exception cref="FileNotFoundException">Thrown if the JSON file is not found.</exception>
    /// <exception cref="JsonException">Thrown if the JSON parsing fails.</exception>
    public static List<string> LoadRwpPointsFromJson(CalibrationManager manager, string jsonFilePath)
    {
        if (!File.Exists(jsonFilePath))
        {
            throw new FileNotFoundException($"RWP Configuration file not found at: {jsonFilePath}");
        }

        string jsonString = File.ReadAllText(jsonFilePath);
        RWPConfiguration config = JsonSerializer.Deserialize<RWPConfiguration>(jsonString) 
            ?? throw new JsonException("Failed to deserialize RWPConfiguration.");

        if (config.Points == null || config.Points.Length == 0)
        {
            throw new InvalidOperationException("RWP configuration contains no points.");
        }

        List<string> orderedPointIds = new List<string>();

        foreach (var rwpData in config.Points)
        {
            // Add point to the CalibrationManager
            manager.AddRealWorldReferencePoint(rwpData.Id, rwpData.ToVector3());
            orderedPointIds.Add(rwpData.Id);
        }

        return orderedPointIds;
    }
}